wa/workloads/speedometer/__init__.py

```diff
@@ -12,57 +12,221 @@
+    s = socket.socket()
+    s.bind(("", 0))
+    port = s.getsockname()[1]
+    s.close()
+    return port
+
+
+class Speedometer(Workload):
+
+    name = "speedometer"
+    package_names = ["org.chromium.chrome", "com.android.chrome"]
+    regex = re.compile('text="(\d+.\d+)" resource-id="result-number"')
+    description = """
+    A workload to execute the speedometer 2.0 web based benchmark
 
     Test description:
-    1. Open chrome
-    2. Navigate to the speedometer website - http://browserbench.org/Speedometer/
-    3. Execute the benchmark
+    1. Host a local copy of the Speedometer website, and make it visible to the device via ADB.
+    2. Open chrome via an intent to access the local copy.
+    3. Execute the benchmark - the copy has been modified to trigger the start of the benchmark.
 
-    known working chrome version 80.0.3987.149
-    '''
+    known working chrome version 83.0.4103.106
+    """
 
     parameters = [
-        Parameter('speedometer_version', allowed_values=versions, kind=str, default='2.0',
-                  description='''
-                  The speedometer version to be used.
-                  ''')
+        Parameter(
+            "chrome_package",
+            allowed_values=package_names,
+            kind=str,
+            default="com.android.chrome",
+            description="""
+                  The app package for the browser that will be launched.
+                  """,
+        ),
     ]
 
-    requires_network = True
-
     def __init__(self, target, **kwargs):
         super(Speedometer, self).__init__(target, **kwargs)
-        self.gui.timeout = 1500
-        self.gui.uiauto_params['version'] = self.speedometer_version
+        self.archive_asset = os.path.join(
+            os.path.dirname(__file__), "speedometer_archive.tgz"
+        )
+
+    def setup(self, context):
+        super(Speedometer, self).setup(context)
+
+        self.document_root = os.path.join(context.output_directory, "document_root")
+
+        self.webserver_port = get_available_webserver_port()
+
+        logging.info(
+            "Using port {} for temporary webserver".format(self.webserver_port)
+        )
+
+        self.speedometer_url = "http://localhost:{}/Speedometer2.0/index.html".format(self.webserver_port)
+
+        # Host a copy of Speedometer locally
+        subprocess.check_output(
+            "tar -xf {} -C {}".format(self.archive_asset, context.output_directory),
+            shell=True,
+        )
+
+        webserver_cmd = "cd {} && python3 -mhttp.server {}".format(
+            self.document_root, self.webserver_port
+        )
+
+        with open(os.devnull, "w") as dev_null:
+            self.webserver = subprocess.Popen(
+                webserver_cmd, shell=True, stdout=dev_null, stderr=dev_null
+            )
+
+        subprocess.check_output(
+            "adb reverse tcp:{0} tcp:{0}".format(self.webserver_port), shell=True
```

Comment:
Do not use `adb` directly -- all interactions must go through devlib. Add an appropriate utility there. Or at least used devlib's `adb_command`.